win32ss/gdi/ntgdi/freetype.c

```diff
@@ -1048,331 +1048,336 @@ WeightFromStyle(const char *style_name)
+                EngSetLastError(ERROR_NOT_ENOUGH_MEMORY);
+                DPRINT1("ERROR_NOT_ENOUGH_MEMORY\n");
+                return 0;   /* failure */
             }
-            ExFreePoolWithTag(PrivateEntry, TAG_FONT);
-        }
-        if (FontGDI->Filename)
-            ExFreePoolWithTag(FontGDI->Filename, GDITAG_PFF);
-        EngFreeMem(FontGDI);
-        SharedFace_Release(SharedFace);
-        ExFreePoolWithTag(Entry, TAG_FONT);
-        return 0;
-    }
 
-    os2_version = 0;
-    IntLockFreeType();
-    pOS2 = (TT_OS2 *)FT_Get_Sfnt_Table(Face, FT_SFNT_OS2);
-    if (pOS2)
-    {
-        os2_version = pOS2->version;
-        os2_ulCodePageRange1 = pOS2->ulCodePageRange1;
-        os2_usWeightClass = pOS2->usWeightClass;
-    }
-    IntUnLockFreeType();
-
-    if (pOS2 && os2_version >= 1)
-    {
-        /* get charset and weight from OS/2 header */
+            /* allocate a FONTGDI */
+            FontGDI = EngAllocMem(FL_ZERO_MEMORY, sizeof(FONTGDI), GDITAG_RFONT);
+            if (!FontGDI)
+            {
+                SharedFace_Release(SharedFace);
+                ExFreePoolWithTag(Entry, TAG_FONT);
+                EngSetLastError(ERROR_NOT_ENOUGH_MEMORY);
+                DPRINT1("ERROR_NOT_ENOUGH_MEMORY\n");
+                return 0;   /* failure */
+            }
 
-        /* Make sure we do not use this pointer anymore */
-        pOS2 = NULL;
+            /* set face */
+            FontGDI->SharedFace = SharedFace;
+            FontGDI->CharSet = ANSI_CHARSET;
+            FontGDI->OriginalItalic = ItalicFromStyle(Face->style_name);
+            FontGDI->RequestItalic = FALSE;
+            FontGDI->OriginalWeight = WeightFromStyle(Face->style_name);
+            FontGDI->RequestWeight = FW_NORMAL;
+
+            /* Entry->FaceName */
+            RtlInitAnsiString(&AnsiString, Face->family_name);
+            RtlAnsiStringToUnicodeString(&Entry->FaceName, &AnsiString, TRUE);
+
+            /* Entry->StyleName */
+            if (Face->style_name && Face->style_name[0] &&
+                strcmp(Face->style_name, "Regular") != 0)
+            {
+                RtlInitAnsiString(&AnsiString, Face->style_name);
+                RtlAnsiStringToUnicodeString(&Entry->StyleName, &AnsiString, TRUE);
+            }
+            else
+            {
+                RtlInitUnicodeString(&Entry->StyleName, NULL);
+            }
 
-        for (BitIndex = 0; BitIndex < MAXTCIINDEX; ++BitIndex)
-        {
-            if (os2_ulCodePageRange1 & (1 << BitIndex))
+            /* FontGDI->CharSet */
+            if (FT_IS_SFNT(Face))
             {
-                if (g_FontTci[BitIndex].ciCharset == DEFAULT_CHARSET)
-                    continue;
+                FontGDI->CharSet = IntGetCharSet(iCharSet, os2_ulCodePageRange1);
 
-                if ((CharSetIndex == -1 && CharSetCount == 0) ||
-                    CharSetIndex == CharSetCount)
+                /* FIXME: CharSet is invalid on Marlett */
```

Comment:
In Windows' Marlett? or in our Marlett? because if it's in our Marlett that the CharSet is invalid, the only correct fix would be to fix our font file.

Reply:
Our Marlett's CharSet is wrong (`DEFAULT_CHARSET`). Windows' Marlett has `SYMBOL_CHARSET` correctly.

Reply:
ok, so since it's our Marlett, we may just as well patch the font file I suppose.